### PR TITLE
BFD-1604 - added the Organization Name to the contained Organization resource for V2

### PR DESCRIPTION
**Pull request by cbrunefearless**
**JIRA Ticket:**
[BFD-1604](https://jira.cms.gov/browse/BFD-1604)

**User Story or Bug Summary:**

User Story:

As a peering partner customer, I want name added to the contained Organization resource, so that I am receiving CARIN compliant data.
AC:

ExplanationOfBenefit.contained[0].ofType(Organization) 
[Organization.name](http://organization.name/): minimum required = 1, but only found 0 (from http://hl7.org/fhir/us/carin-bb/StructureDefinition/C4BB-Organization)
CARIN Confluence Page - https://confluence.cms.gov/display/BFD/CARIN+Conformance+-+Eob

---

### What Does This PR Do?

Adds org name to v2.  Also has the same functionality for v1 but we arent populating v1 according to Brady

### What Should Reviewers Watch For?

Make sure name is added to the org resource
If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [X] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [X]  No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [X]  Yes
      * [ ] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [X]  No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [X]  I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [X]  This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [X]  This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X]  All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X]  All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [X]  Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-server/bfd-server-war/src/test/resources/endpoint-responses/v2/eobReadCarrierWithTaxNumbers.json [hunk 1]
```diff
@@ -227,9 +227,8 @@
             "display" : "National Provider Identifier"
           } ]
         },
-        "value" : "1497758544"
-      },
-      "display" : "CUMBERLAND COUNTY HOSPITAL SYSTEM, INC"
```
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/test/resources/endpoint-responses/v2/eobReadCarrierWithTaxNumbers.json:232]: why did we lose this?

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/commons/NPIOrgDataUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+
+import java.util.Optional;
+import org.junit.jupiter.api.Test;
+
+public class NPIOrgDataUtilsTest {
+
+  /** Return Fake NPI Org Data when parameter is true */
```
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/commons/NPIOrgDataUtilsTest.java:11]: which parameter is true? these comments and test names don't seem to line up with the test code

**Review by christrotteradhoc — CHANGES_REQUESTED** [review]
Great work getting this implemented Chris. I think there are a few small things to take care of that I have made comments on (those things may apply to several places in the code) and there is a larger question in my mind about the way that we locate the file.

**Comment by christrotteradhoc** [comment]
Few other comments that are not related to any particular piece of code -- The PR title has brackets around the BFD number and shouldn't, the PR title is not descriptive enough, and the PR description needs a bit of cleanup. We are also missing the api_changelog entry I believe.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/NPIDataUtilityApp.java [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+ * converts it to UTF-8 format.
+ *
+ * <p>See the <code>download-fda-drug-data</code> execution of <code>exec-maven-plugin</code> in
+ * this project's <code>pom.xml</code> for details on how this utility is run during the project's
+ * build.
+ */
+public final class NPIDataUtilityApp {
+  /**
+   * The name of the classpath resource (for the project's main web application) for the FDA
+   * "Products" TSV file.
```
[the remark this answers is not in the document]
Comment (cbrunefearless) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/NPIDataUtilityApp.java:43]: done

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/NPIDataUtilityApp.java [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+          }
+        } finally {
+          // Recursively delete the working dir.
+          Files.walk(workingDir)
+              .sorted(Comparator.reverseOrder())
+              .map(Path::toFile)
+              .peek(System.out::println)
+              .forEach(File::delete);
+        }
+      } catch (Exception exception) {
```
[the remark this answers is not in the document]
Comment (cbrunefearless) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/NPIDataUtilityApp.java:105]: done

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/SpringConfiguration.java [hunk 5]
```diff
@@ -75,6 +76,15 @@ public class SpringConfiguration {
    */
   public static final String PROP_INCLUDE_FAKE_DRUG_CODE = "bfdServer.include.fake.drug.code";
 
+  /**
+   * The {@link String } Boolean property that is used to enable the fake npi number (0000000000)
+   * that is used for integration testing. When this property is set to the string 'true', this fake
+   * npi number will be appended to the npi org data lookup map to avoid test failures that result
```
[the remark this answers is not in the document]
Comment (cbrunefearless) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/SpringConfiguration.java:82]: done

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/utils/EndpointJsonResponseComparatorV2IT.java [hunk 6]
```diff
@@ -219,6 +219,7 @@ public void verifyCorrectEndpointResponse(String endpointId, Supplier<String> en
    * Generates the "golden" files, i.e. the approved responses to compare to. Run by commenting out
    * the <code>@Ignore</code> annotation and running this method as JUnit.
    */
+  //
```
[the remark this answers is not in the document]
Comment (cbrunefearless) [reply apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/utils/EndpointJsonResponseComparatorV2IT.java:222]: done

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/OutpatientClaimTransformerV2.java [hunk 7]
```diff
@@ -50,7 +50,10 @@ static ExplanationOfBenefit transform(TransformerContext transformerContext, Obj
     }
 
     ExplanationOfBenefit eob =
-        transformClaim((OutpatientClaim) claim, transformerContext.getDrugCodeDisplayLookup());
+        transformClaim(
+            (OutpatientClaim) claim,
+            transformerContext.getDrugCodeDisplayLookup(),
```
[the remark this answers is not in the document]
Comment (cbrunefearless) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/OutpatientClaimTransformerV2.java:55]: done

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/NPIDataUtilityApp.java [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+        new BufferedOutputStream(new FileOutputStream(filePath.toFile().getAbsolutePath()));
+    byte[] bytesIn = new byte[BUFFER_SIZE];
+    int read = 0;
+    while ((read = zipIn.read(bytesIn)) != -1) {
+      bos.write(bytesIn, 0, read);
+    }
+    bos.close();
+  }
+
+  private static String getFileName(boolean getMonthBefore) {
```
[the remark this answers is not in the document]
Comment (cbrunefearless) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/NPIDataUtilityApp.java:227]: done

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/NPIOrgDataLookup.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+  public static final String FAKE_NPI_ORG_NAME = "Fake ORG Name";
+
+  /** */
+  private final Map<String, String> npiOrgHashMap = new HashMap<>();
+
+  /** */
+  private static NPIOrgDataLookup npiOrgLookupForTesting;
+
+  /** */
+  private static NPIOrgDataLookup npiOrgLookupForProduction;
```
[the remark this answers is not in the document]
Comment (cbrunefearless) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/NPIOrgDataLookup.java:32]: done

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/NPIDataUtilityApp.java [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+    if (!Files.exists(convertedNpiDataFile)) {
+      String fileName = getFileName(false);
+      try {
+        try {
+          buildNPIResource(convertedNpiDataFile, workingDir, fileName);
+        } catch (FileNotFoundException ex) {
+          try {
+            fileName = getFileName(true);
+            buildNPIResource(convertedNpiDataFile, workingDir, fileName);
+          } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (cbrunefearless) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/NPIDataUtilityApp.java:94]: done

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/NPIOrgDataLookup.java [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+      // skip first line which is header
+      npiOrgIn.readLine();
+      while ((line = npiOrgIn.readLine()) != null) {
+        String npiProductColumns[] = line.split("\t");
+        try {
+          npiOrgHashMap.put(
+              npiProductColumns[0].replace("\"", ""), npiProductColumns[1].replace("\"", ""));
+        } catch (StringIndexOutOfBoundsException e) {
+          continue;
+        }
```
[the remark this answers is not in the document]
Comment (cbrunefearless) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/NPIOrgDataLookup.java:119]: put the replace in the etl..

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/NPIDataUtilityApp.java [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+    }
+
+    // unzip NPI file
+    unzip(downloadedNpiZipFile, workingDir);
+    File f = new File(workingDir.toString());
+    File[] matchingFiles =
+        f.listFiles(
+            new FilenameFilter() {
+              public boolean accept(File dir, String name) {
+                return name.startsWith("npidata_pfile_") && !name.endsWith("_FileHeader.csv");
```
[the remark this answers is not in the document]
Comment (christrotteradhoc) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/NPIDataUtilityApp.java:136]: also please handle the condition where more than one files match (an error condition probably)

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/DataUtilityCommons.java [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+import org.apache.commons.io.FileUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Commons library for upzipping files */
+public final class DataUtilityCommons {
```
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/DataUtilityCommons.java:29]: add private constructor for utility class

File: apps/bfd-server/bfd-server-war/src/test/resources/endpoint-responses/v2/eobReadCarrierWithTaxNumbers.json (hunk 1 above)
Comment (cbrunefearless) [reply apps/bfd-server/bfd-server-war/src/test/resources/endpoint-responses/v2/eobReadCarrierWithTaxNumbers.json:232]: we added a different value

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/DataUtilityCommons.java (hunk 13 above)
Comment (cbrunefearless) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/DataUtilityCommons.java:29]: added

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/NPIDataUtilityApp.java [hunk 14]
```diff
@@ -0,0 +1,274 @@
...
+import org.apache.commons.io.FileUtils;
+
+/**
+ * A simple application that downloads the FDA NDC (national drug code) file; unzips it and then
+ * converts it to UTF-8 format.
+ *
+ * <p>See the <code>download-fda-drug-data</code> execution of <code>exec-maven-plugin</code> in
+ * this project's <code>pom.xml</code> for details on how this utility is run during the project's
+ * build.
+ */
```
[the remark this answers is not in the document]
Comment (cbrunefearless) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/NPIDataUtilityApp.java]: fixed

**Comment by cbrunefearless** [comment]
> Great work getting this implemented Chris. I think there are a few small things to take care of that I have made comments on (those things may apply to several places in the code) and there is a larger question in my mind about the way that we locate the file.

thats fine

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/NPIDataUtilityApp.java [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+            .onMalformedInput(CodingErrorAction.REPORT)
+            .onUnmappableCharacter(CodingErrorAction.REPORT);
+
+    try (FileInputStream is = new FileInputStream(originalNpiDataFile.toFile().getAbsolutePath());
+        BufferedReader reader = new BufferedReader(new InputStreamReader(is, inDec));
+        FileOutputStream fw =
+            new FileOutputStream(convertedNpiDataFile.toFile().getAbsolutePath());
+        BufferedWriter out = new BufferedWriter(new OutputStreamWriter(fw, outEnc))) {
+
+      // TODO: Explain file format and show example
```
Comment (mackec1) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/NPIDataUtilityApp.java:155]: For new code such as this, aren't we trying to minimize/elimnated TODOs?

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/NPIOrgDataLookup.java [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+  }
+
+  /**
+   * Retrieves the Org Data for NPI Display for the given NPI number using the npi file downloaded
+   * during the build.
+   *
+   * @param npiNumber - npiNumber value in claim records
+   * @return the npi org data display string
+   */
+  public Optional<String> retrieveNPIOrgDisplay(Optional<String> npiNumber) {
```
Comment (mackec1) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/NPIOrgDataLookup.java:78]: This seems kinda wierd; the signature for the method returns an Optional. The caller will expect an Optional and will probably be testing for .isPresent; so returning a null will probably result in NPE. Shouldn't we just be rturning an Optional.empty() if we don't get a hit?

File: apps/bfd-server/bfd-server-war/src/test/resources/endpoint-responses/v1/eobReadOutpatient.json [hunk 17]
```diff
@@ -147,9 +146,8 @@
     } ],
     "identifier" : {
       "system" : "http://hl7.org/fhir/sid/us-npi",
-      "value" : "1497758544"
-    },
-    "display" : "CUMBERLAND COUNTY HOSPITAL SYSTEM, INC"
+      "value" : "0000000000"
+    }
```
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/test/resources/endpoint-responses/v1/eobReadOutpatient.json:150]: I think this was only supposed to be for v2, right?

File: apps/bfd-server/dev/api-changelog.md [hunk 18]
```diff
@@ -1,5 +1,38 @@
 # API Changelog
 
+## BFD-1604: added the Organization Name to the contained Organization resource for V2
```
Comment (christrotteradhoc) [inline apps/bfd-server/dev/api-changelog.md:3]: BFD-1604 is showing up as closed due to being a duplicate

File: apps/bfd-server/bfd-server-war/src/test/resources/endpoint-responses/v2/eobReadCarrierWithTaxNumbers.json (hunk 1 above)
Comment (christrotteradhoc) [reply apps/bfd-server/bfd-server-war/src/test/resources/endpoint-responses/v2/eobReadCarrierWithTaxNumbers.json:232]: this seems like a backward incompatibility

**Review by christrotteradhoc — COMMENTED** [review]
I've not gone back through all of the code yet but I think there are a few high level things that we need to get sorted out -- I'm not sure we ever actually groomed the ticket and determined the API AC. Can we bring this to the next grooming session?

**Pull request closed without merging**